File: java/src/main/java/BookListingFormatter.java
import static java.lang.String.format;

public class BookListingFormatter {

    public static String createListingText(Book book, String author, Integer starRating, String bookPrize, String authorPrize) {
        StringBuilder stringBuilder = new StringBuilder();
        if(bookPrize != null && !bookPrize.isEmpty()){
            stringBuilder.append(bookPrize);
            stringBuilder.append(" winner - ");
        }
        stringBuilder.append(format("%s by %s", book.getTitle(), author));

        if(authorPrize != null && !authorPrize.isEmpty()){
            stringBuilder.append(" (Winner of ");
            stringBuilder.append(authorPrize);
            stringBuilder.append(")");
        }

        if(starRating != null){
            if(1 == starRating) { 
                stringBuilder.append(": 1 star");
            } else {
                stringBuilder.append(": ");
                stringBuilder.append(starRating);
                stringBuilder.append(" stars");
            }
        }
        return stringBuilder.toString();
    }

}

File: typescript/src/BookListingFormatter.ts
import { Book } from "./Book";

export class BookListingFormatter {
    public static createListingText(book: Book, author: string, starRating: number | null, bookPrize: string | null, authorPrize: string | null): string {
      let listing = '';
  
      if (bookPrize !== null && bookPrize.length > 0) {
        listing += `${bookPrize} winner - `;
      }
  
      listing += `${book.getTitle()} by ${author}`;
  
      if (authorPrize !== null && authorPrize.length > 0) {
        listing += ` (Winner of ${authorPrize})`;
      }
  
      if (starRating !== null) {
        if (starRating === 1) {
          listing += ': 1 star';
        } else {
          listing += `: ${starRating} stars`;
        }
      }
  
      return listing;
    }
  }
  